flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/NetworkBufferPool.java

```diff
@@ -170,9 +194,16 @@ public void recycle(MemorySegment segment) {
 					throw new IllegalStateException("Buffer pool is destroyed.");
 				}
 
-				final MemorySegment segment = availableMemorySegments.poll(2, TimeUnit.SECONDS);
-				if (segment != null) {
-					segments.add(segment);
+				synchronized (availabilityHelper) {
+					final MemorySegment segment = availableMemorySegments.poll();
```

Comment:
We could deduplicate some common codes with above `#requestMemorySegment()`.
`
@Nullable
private MemorySegment internalRequestMemorySegment() {
		assert Thread.holdsLock(availableMemorySegments);

		final MemorySegment segment = availableMemorySegments.poll();
		if (availableMemorySegments.isEmpty() && segment != null) {
			availabilityHelper.resetUnavailable();
		}
		return segment;
}
`
Then we could refactor here as 
`
final MemorySegment segment = internalRequestMemorySegment();
if (segment == null) {
	availabilityHelper.wait(2000);
}
`